subprojects/core/src/main/java/org/gradle/api/internal/file/collections/ImmutableFileCollection.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.file.collections;
+
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.internal.file.AbstractFileCollection;
+import org.gradle.api.internal.file.FileCollectionInternal;
+import org.gradle.api.internal.file.FileResolver;
+import org.gradle.api.internal.file.IdentityFileResolver;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+public abstract class ImmutableFileCollection extends AbstractFileCollection {
```

Comment:
❌ Shouldn't immutable file collections be able to convey information about build dependencies, too? So for example if an output `RegularFileProperty` is added to `ResolvingImmutableFileCollection`, then the file collection would also depend on the task which creates the `RegularFileProperty`?

Reply:
I added this to the Crucible review: https://code-review.gradle.org/cru/REVIEW-6580#c39576

Reply:
Thank you!